michbl/assignment2/BinaryTree.hpp

```diff
@@ -0,0 +1,194 @@
+struct Node {
+    T key;
+    Node<T> *left = nullptr;
+    Node<T> *right = nullptr;
+};
+
+template <typename T>
+class BinaryTree {
+private:
+    Node<T> *root;
+
+    void copyTree(Node<T> *parent, Node<T> *newNode, string side) {
+        if (newNode == nullptr) {
+            return;
+        }
+
+        Node<T> *child = new Node<T>;
+        child->key = newNode->key;
+        if (side == "left") {
+            parent->left = child;
+        }
+        else {
+            parent->right = child;
+        }
+        copyTree(child, newNode->left, "left");
+        copyTree(child, newNode->right, "right");
+    }
+
+    void insert(Node<T> *node, T key, T parentKey, bool &success) {
+        if (node == nullptr) {
+            return;
+        }
+        if (node->key == parentKey) {
+            if (node->left == nullptr) {
+                Node<T> *child = new Node<T>;
+                child->key = key;
+                node->left = child;
+                success = true;
+                return;
+            }
+            else if (node->right == nullptr) {
+                Node<T> *child = new Node<T>;
+                child->key = key;
+                node->right = child;
+                success = true;
+                return;
+            }
+            else { // parent already has 2 children
+                return;
+            }
+        }
+        insert(node->left, key, parentKey, success);
+        if (success) {
+            return;
+        }
+        insert(node->right, key, parentKey, success);
+    }
+
+    void destroyTree(Node<T> *node) {
+        if (node == nullptr) {
+            return;
+        }
+        destroyTree(node->left);
+        destroyTree(node->right);
+        delete node;
+    }
+
+    void printTree(Node<T> *node) {
+        if (node == nullptr) {
+            return;
+        }
+        printTree(node->left);
+        printTree(node->right);
+        cout << node->key << " ";
+    }
+
+    void printAncestors(Node<T> *node, T key, bool &found) {
+        if (node == nullptr) {
+            return;
+        }
+        if (node->key == key) {
+            found = true;
+            return;
+        }
+        if (!found) {
+            printAncestors(node->left, key, found);
+        }
+        if (!found) {
+            printAncestors(node->right, key, found);
+        }
+        if (found) {
+            if (node == this->root) {
+                cout << node->key << endl;
+            }
+            else {
+                cout << node->key << ", ";
+            }
+        }
+    }
+
+    bool bothKeysAreInSubtree(Node<T> *node, T key1, T key2) {
+        bool found1 = false, found2 = false;
+        bothKeysAreInSubtreeAux(node->left, key1, key2, found1, found2);
+        bothKeysAreInSubtreeAux(node->right, key1, key2, found1, found2);
+        return found1 && found2;
+    }
+
+    void bothKeysAreInSubtreeAux(Node<T> *node, T key1, T key2, bool &found1, bool &found2) {

```

Comment:
Tip:
You can use an additional helper function which checks if only one key is in a subtree, then `bothKeysAreInSubtree` function can call this function twice, once for each key.
I bet that it can make the code a little shorter (blind bet, I may be wrong).
Also, I think that a function for a single key is more valuable in terms of it can be used out of this context.